pym/portage/meta-manifest.py

```diff
@@ -0,0 +1,276 @@
+                        self.fhashdict[ftype] = {}
+		for pkg_dir in os.listdir(catdir):
+			pkg_dir = os.path.join(catdir, pkg_dir)
+			if os.path.isdir(pkg_dir):
+				for file in os.listdir(pkg_dir):
+					fpath = os.path.join(pkg_dir, file)
+					if os.path.isfile(fpath) and fpath.endswith('Manifest'):
+						new_mf = MetaManifest(fpath.replace('Manifest', ""))
+						new_mf.verify_manifest()
+						f = fpath.replace(self.pkgdir, "")
+						self.fhashdict["MANIFEST"][f] = perform_multiple_checksums(fpath, self.hashes)
+
+	def create_profile(self):
+		'''Creates a MetaManifest file in the profiles directory'''
+		profiledir = self.pkgdir
+		for ftype in MANIFEST2_IDENTIFIERS:
+                        self.fhashdict[ftype] = {}
+		for profiledir, profiledir_dir, files in os.walk(profiledir):
+			for f in files:
+				try:
+					f = _unicode_decode(f,encoding=_encodings['fs'], errors='strict')
+					profiledir = _unicode_decode(profiledir,encoding=_encodings['fs'], errors='strict')
+				except UnicodeDecodeError:
+					continue
+				fpath = os.path.join(profiledir, f)
+				ftype = guessManifestFileType(fpath, is_pkg=False)
+				if ftype == "OTHER":
+					ftype = "DATA"
+				f = fpath.replace(self.pkgdir, "")
+				if not f.endswith("Manifest"):
+					self.fhashdict[ftype][f] = perform_multiple_checksums(fpath, self.hashes)
+
+	def checkFileHashes(self, ftype, fname, ignoreMissing=False, hash_filter=None):
+                digests = _filter_unaccelarated_hashes(self.fhashdict[ftype][fname])
+                if hash_filter is not None:
+                        digests = _apply_hash_filter(digests, hash_filter)
+                try:
+                        if ftype not in ('DIST'):
+                                ok, reason = verify_all(self._getAbsname(ftype, fname), digests)
+                                if not ok:
+                                        raise DigestException(tuple([self._getAbsname(ftype, fname)]+list(reason)))
+                                return ok, reason
+                except FileNotFound as e:
+                        if not ignoreMissing:
+                                raise
+                        return False, _("File Not Found: '%s'") % str(e)
+
+
+	def validate_signature(self):
+		'''Validate the Manifest's file signature.'''
+		self.gkeys = GkeysInterface('portage', '/')
+		is_good, has_sig = self.gkeys.verify_file(self.getFullname())
+		return (is_good, has_sig)
+
+	def verify_manifest(self, verify_pkgs = True):
+		'''Verifies the manifest's validity'''
+		mfdir = self.pkgdir
+		mftype = self.find_repotype(mfdir)
+		if mftype == 'root':
+			for ftype in MANIFEST2_IDENTIFIERS:
+				for f in self.fhashdict[ftype]:
+					if ftype != 'MANIFEST':
+						self.checkFileHashes(ftype, f, ignoreMissing=False)
+					else:
+						mfpath = os.path.join(mfdir, f.replace("Manifest", ""))
+						missing = False
+						if not os.path.exists(mfpath) and self.find_repotype(f) is 'category':
```

Comment:
Do we really have to call this `find_repotype` method in the loop? If the result is the same for each iteration of the loop, we should call it before the loop.